src/bottom-bar/use-on-complete-callback.js

```diff
@@ -0,0 +1,166 @@
+                        ) {
+                            setRightHandPanel(validationResultsSidebarId)
+                        }
+                    }
+                )
+                .catch((e) => {
+                    console.error(e)
+                    // For now this can fail "silently".
+                    // ES6 Promises don't have a mechanism for reporting
+                    // multiple promise failures and we def. want to notify the
+                    // user when form completion fails as well.
+                    // It should be enough to show a warning to the user
+                    showWarningAlert(validationFailedWarningMessage)
+                }),
+            setFormCompletion({ completed: true }).catch(() => {
+                throw new Error(completingFormFailedMessage)
+            }),
+        ])
+    }
+}
+
+function useOnCompleteWithoutValidationClick() {
+    const { mutateAsync: setFormCompletion } = useSetFormCompletionMutation()
+    return () => setFormCompletion({ completed: true })
+}
+
+export default function useOnCompleteCallback() {
+    const { offline } = useConnectionStatus()
+    const { data: metadata } = useMetadata()
+    const [dataSetId] = useDataSetId()
+    const dataSet = selectors.getDataSetById(metadata, dataSetId)
+    const { validCompleteOnly } = dataSet
+    const { show: showErrorAlert } = useAlert((message) => message, {
+        critical: true,
+    })
+    const setFormCompletionMutationKey = useSetFormCompletionMutationKey()
+    const isLoading = useIsMutating(setFormCompletionMutationKey)
+    const cancelCompletionMutation = useImperativeCancelCompletionMutation()
+    const onCompleteWhenValidRequiredClick =
+        useOnCompleteWhenValidRequiredClick()
+    const onCompleteWhenValidNotRequiredClick =
+        useOnCompleteWhenValidNotRequiredClick()
+    const onCompleteWithoutValidationClick =
+        useOnCompleteWithoutValidationClick()
+
+    return () => {
+        if (isLoading) {
+            cancelCompletionMutation()
+        }
+
+        let promise
+        if (isLoading && offline) {
+            // No need to complete when the completion request
+            // hasn't been sent yet due to being offline.
+            // It's important to still perform the request online as we don't
+            // know if the mutation actually reached the server already
+            return Promise.resolve()
+        } else if (offline) {
+            // When offline, we can't validate, so we simply complete the form
+            promise = onCompleteWithoutValidationClick()
+        } else if (validCompleteOnly && !offline) {
+            promise = onCompleteWhenValidRequiredClick()
+        } else {
+            promise = onCompleteWhenValidNotRequiredClick()
+        }
+
+        return (
+            promise
+                // this will eventually catch any error thrown and display the error message
+                .catch((e) => showErrorAlert(e.message))
```

Comment:
I think this should be handled on `onError` to `useMutation`. 

We could also look into having a global handler for this.

Reply:
Let's do this in another PR so we can get this merged today

Reply:
https://dhis2.atlassian.net/browse/TECH-1434